src/frontend/src/handler/privilege.rs

```diff
@@ -62,79 +66,41 @@ pub(crate) fn resolve_relation_privileges(
             if let crate::binder::BoundSetExpr::Select(select) = &query.query.body {
                 if let Some(sub_relation) = &select.from {
-                    resolve_relation_privileges(sub_relation, mode, objects);
+                    resolve_relation_privileges(sub_relation, mode, objects, databases);
                 }
             }
         }
         Relation::Join(join) => {
-            resolve_relation_privileges(&join.left, mode, objects);
-            resolve_relation_privileges(&join.right, mode, objects);
+            resolve_relation_privileges(&join.left, mode, objects, databases);
+            resolve_relation_privileges(&join.right, mode, objects, databases);
         }
         Relation::WindowTableFunction(table) => {
-            resolve_relation_privileges(&table.input, mode, objects)
+            resolve_relation_privileges(&table.input, mode, objects, databases);
         }
         _ => {}
     };
 }
 
-/// resolve privileges in `stmt`
-pub(crate) fn resolve_privileges(stmt: &BoundStatement) -> Vec<ObjectCheckItem> {
-    let mut objects = Vec::new();
-    match stmt {
-        BoundStatement::Insert(ref insert) => {
-            let object = ObjectCheckItem {
-                owner: insert.owner,
-                mode: AclMode::Insert,
-                object: PbObject::TableId(insert.table_id.table_id),
-            };
-            objects.push(object);
-            if let crate::binder::BoundSetExpr::Select(select) = &insert.source.body {
-                if let Some(sub_relation) = &select.from {
-                    resolve_relation_privileges(sub_relation, AclMode::Select, &mut objects);
-                }
-            }
-        }
-        BoundStatement::Delete(ref delete) => {
-            let object = ObjectCheckItem {
-                owner: delete.owner,
-                mode: AclMode::Delete,
-                object: PbObject::TableId(delete.table_id.table_id),
-            };
-            objects.push(object);
-        }
-        BoundStatement::Update(ref update) => {
-            let object = ObjectCheckItem {
-                owner: update.owner,
-                mode: AclMode::Update,
-                object: PbObject::TableId(update.table_id.table_id),
-            };
-            objects.push(object);
-        }
-        BoundStatement::Query(ref query) => objects.extend(resolve_query_privileges(query)),
-        BoundStatement::DeclareCursor(ref declare_cursor) => {
-            objects.extend(resolve_query_privileges(&declare_cursor.query))
-        }
-        BoundStatement::FetchCursor(_) => unimplemented!(),
-        BoundStatement::CreateView(ref create_view) => {
-            objects.extend(resolve_query_privileges(&create_view.query))
-        }
-    };
-    objects
-}
-
 /// resolve privileges in `query`
-pub(crate) fn resolve_query_privileges(query: &BoundQuery) -> Vec<ObjectCheckItem> {
+fn resolve_query_privileges(query: &BoundQuery) -> (Vec<ObjectCheckItem>, HashSet<DatabaseId>) {
```

Comment:
It seems that all usages are first collect the two variables, and then extend to the outer list or sets. If so, we may pass the &mut of the outer list or sets here instead of returning a new one.